soccer/gameplay/evaluation/double_touch.py

```diff
@@ -33,66 +36,105 @@ def __init__(self):
         self.add_transition(
             DoubleTouchTracker.State.start,
             DoubleTouchTracker.State.restart_play_began,
-            lambda: (main.root_play().play != None and main.root_play().play.__class__.is_restart() and main.game_state().is_our_restart()),
+            lambda: (main.root_play().play is not None and
+                     main.root_play().play.__class__.is_restart() and
+                     main.game_state().is_our_restart()),
             'we start running an offensive restart play')
 
         self.add_transition(
             DoubleTouchTracker.State.restart_play_began,
             DoubleTouchTracker.State.kicking,
-            lambda: (any(bot.has_ball() for bot in main.our_robots())) and not main.game_state().is_placement(),
-            'one of our bots has the ball')
+            lambda: ((any(self._has_ball(bot) for bot in main.our_robots())) or
+                     main.game_state().is_playing()) and not main.game_state().is_placement(),
+            'one of our bots has the ball or the ball was kicked')
 
         self.add_transition(DoubleTouchTracker.State.kicking,
                             DoubleTouchTracker.State.kicker_forbidden,
-                            lambda: not self.kicker_has_possession(),
-                            'kicker kicks or fumbles ball')
+                            # The ball is no longer in restart, we have begun playing
+                            (lambda: main.game_state().is_playing() or
```

Comment:
Not sure about rules, but doesn't our robot also have to lose control of the ball? Will the play state always change?

Reply:
Since we always have a line kick for restarts, as soon as we exit the restart state it means we touched the ball. For pivot kick though this is probably broken and will interrupt while the bot is kicking....